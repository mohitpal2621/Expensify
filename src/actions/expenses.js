import { v4 as uuid } from "uuid";

// ADD_EXPENSE action generator
export const addExpense = ({ description='', note='', amount=0, createdAt=0 }) => ({
    type: 'ADDEXPENSE',
    expense: {
        id: uuid(),
        description,
        note,
        amount,
        createdAt
    }
});

// REMOVE_EXPENSE action generator
export const removeExpense = (id) => ({
    type: 'REMOVEEXPENSE',
    id
});

// EDIT_EXPENSE action generator
export const editExpense = (id, updates) => ({
        type: 'EDITEXPENSE',
        id,
        updates
});