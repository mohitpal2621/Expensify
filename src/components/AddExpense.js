import React from "react";

const AddExpense = () => {
    return (
        <div>
            This is from my AddExpense comp.
        </div>
    )
};

export default AddExpense;