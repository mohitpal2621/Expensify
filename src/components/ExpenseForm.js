import React from "react";
import moment from "moment";
import 'react-dates/initialize'
import { SingleDatePicker } from "react-dates";
import 'react-dates/lib/css/_datepicker.css';

export default class ExpenseForm extends React.Component {
    state = {
        description: '',
        note: '',
        amount: '',
        createdAt: moment(),
        calendarFocused: false,
        error: ''
    }

    onDescChange = (e) => {
        this.setState(() => (
            {description: e.target.value}
        ));
    }

    onNoteChange = (e) => {
        this.setState(() => (
            {note: e.target.value}
        ));
    } 

    onAmountChange = (e) => {
        const amt = e.target.value;
        if(!amt || /^\d+(\.\d{0,2})?$/.test(amt))
            this.setState(() => ( {amount: e.target.value} ));
    }

    onDateChange = (createdAt) => {
        if(createdAt)
            this.setState(() => ({createdAt}));
    }

    onFocusChange= ({focused}) => {
        this.setState(() => ({calendarFocused: focused}))
    }

    onSubmit = (e) => {
        e.preventDefault();
        if(!this.state.description || !this.state.amount){
            this.setState(() => (
                {error: "Please provide description and amount."}
            ));
        }else{
            this.setState(() => ({error: ''}));
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10).toFixed(2),
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note
            });
        }

    }

    render() {
        return (
            <div>
                <form onSubmit={this.onSubmit}>
                    <input 
                        type="text"
                        placeholder="Description"
                        autoFocus
                        value={this.state.description}
                        onChange={this.onDescChange}
                    />
                    <input 
                        type="text"
                        value={this.state.amount}
                        placeholder="Amount"
                        onChange={this.onAmountChange}
                    />
                    <SingleDatePicker
                        date={this.state.createdAt}
                        focused={this.state.calendarFocused}
                        onDateChange={this.onDateChange}
                        onFocusChange={this.onFocusChange}
                        numberOfMonths={1}
                        isOutsideRange={() => false}
                    />
                    <textarea 
                        placeholder="Add a note for your expense (optional)"
                        value={this.state.note}
                        onChange={this.onNoteChange}
                    >
                    </textarea>
                    {this.state.error && <p>{this.state.error}</p>}
                    <button>Add Expense</button>
                </form>
            </div>
        );
    }
}