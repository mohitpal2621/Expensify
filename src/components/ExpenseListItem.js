import React from "react";
import moment from "moment";
import { Link } from "react-router-dom";

const curr = Intl.NumberFormat("en-IN", {
    style: "currency",           //Decimal, percent or currency
    currency: "INR",             //ISO code for countries' currencies
    currencyDisplay: "symbol",   //symbol, code, name
    useGrouping: "auto",
    minimumIntegerDigits: 1,
    maximumFractionDigits: 10,
    minimumFractionDigits: 2,
});

const ExpenseListItem = ({ id, description, amount, createdAt }) => (
    <div>
        <Link to={`/edit/${id}`}>
            <h3>{description}</h3>
        </Link>
        <p>
            {curr.format(amount)} 
            - 
            {moment(createdAt).format('MMMM Do, YYYY')}
        </p>
    </div>
)

export default ExpenseListItem;