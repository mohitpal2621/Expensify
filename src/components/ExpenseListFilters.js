import React from "react";
import { connect } from "react-redux";
import { setEndDate, setStartDate, setTextFilter, sortByAmount, sortByDate } from "../actions/filters";
import { DateRangePicker } from "react-dates";

// If the relevant part of the state the HOC component cares about (in this case, state.filters.text) hasn't changed, 
// the component won't re-render.
class ExpenseListFilters extends React.Component {
    state = {
        calendarFocused: null
    }

    onDatesChange = ({startDate, endDate}) => {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    }

    onFocusChange = (calendarFocused) => {
        this.setState(() => ({calendarFocused}));
    }

    onTextChange = (e) => {
        this.props.setTextFilter(e.target.value);
    }

    onSortChange = (e) => {
        if(e.target.value === 'date')
            this.props.sortByDate();
        else if(e.target.value === 'amount')
            this.props.sortByAmount();
    }

    render(){
        return (
            <div className="content-container">
                <div className="input-group">
                    <div className="input-group__item">
                        <input 
                            className="text-input"
                            type="text" 
                            placeholder="Select expenses"
                            value={this.props.filters.text}
                            onChange={this.onTextChange}
                        />
                    </div>
                     <div className="input-group__item">
                        <select 
                            className="select"
                            value={this.props.filters.sortBy}
                            onChange={this.onSortChange}>
                            <option value="date">Date</option>
                            <option value="amount">Amount</option>
                        </select>
                    </div>
                    <div className="input-group__item">
                        <DateRangePicker
                            startDate={this.props.filters.startDate}
                            startDateId="myDateId"
                            endDate={this.props.filters.endDate}
                            endDateId="myDateId"
                            onDatesChange={this.onDatesChange}
                            focusedInput={this.state.calendarFocused}
                            onFocusChange={this.onFocusChange}
                            isOutsideRange={() => false}
                            numberOfMonths={1}
                            showClearDates={true}
                        />
                    </div>
                </div>
            </div> 
        ); 
    }
};

const mapStateToProps = (state) => ({
    filters: state.filters
});

const mapDispatchToProps = (dispatch) => ({
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate)),
    setTextFilter: (val) => dispatch(setTextFilter(val)),
    sortByDate: () => dispatch(sortByDate()),
    sortByAmount: () => dispatch(sortByAmount())
});
  
export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);