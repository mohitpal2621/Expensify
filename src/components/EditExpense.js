// Using Hooks instead of connect()
import React from "react";
import { useDispatch, useSelector } from "react-redux";
import { useNavigate, useParams } from "react-router-dom";
import ExpenseForm from "./ExpenseForm";
import { startEditExpense, startRemoveExpense } from "../actions/expenses";

export const EditExpense = () => {
    const { id } = useParams();
    const dispatch = useDispatch();
    const nav = useNavigate();
    const exp = useSelector(state => state.expenses.find((exp) => exp.id === id));
    
    return (
        <div>
            <div className="page-header">
                <div className="content-container">
                    <h1 className="page-header__title">Edit Expense</h1>
                </div>
            </div>
            <div className="content-container">
                <ExpenseForm
                    expense={exp}
                    onSubmit={(exp) => {
                        dispatch(startEditExpense(id, exp));
                        nav("/");
                    }}
                />
                <button className="button button--secondary" onClick={() => {
                        dispatch(startRemoveExpense(id));
                        nav("/");
                    }}
                >
                    Remove
                </button>
            </div>
        </div>
    );
};

export default EditExpense;