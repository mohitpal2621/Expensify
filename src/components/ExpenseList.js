import React from 'react';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expensesViewer';
import { connect } from 'react-redux';

export const ExpenseList = (props) => (
    <div>
        <h1>This is the Expense List</h1>
        {
            props.expenses.length === 0 ? (
                <p>No Expense</p>
            ) : (
                props.expenses.map((exp) => (
                    <ExpenseListItem key={exp.id} {...exp} />
                ))
            )
        }
        
    </div>
);

const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    }
};

export default connect(mapStateToProps)(ExpenseList);
