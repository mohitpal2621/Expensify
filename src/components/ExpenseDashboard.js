import React from "react";
import ExpenseListFilters from "./ExpenseListFilters";
import ExpenseList from "./ExpenseList";

const ExpenseDashboardPage = () => {
    return (
        <div>
            <ExpenseListFilters/>
            <ExpenseList/>
        </div>
    )
};


export default ExpenseDashboardPage;