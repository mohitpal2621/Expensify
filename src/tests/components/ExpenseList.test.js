import React from 'react';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import {ExpenseList} from '../../components/ExpenseList';
import expenses from '../fixtures/expenses';
import { MemoryRouter } from 'react-router-dom';

test("Should render expense list with expenses", () => {
    const {container} = render(<MemoryRouter><ExpenseList expenses={expenses}/></MemoryRouter>);
    // Assert that the ExpenseList component is rendered
    expect(container).toMatchSnapshot();
});

test("Should render expense list with empty expenses", () => {
    const {container} = render(<MemoryRouter><ExpenseList expenses={[]}/></MemoryRouter>);
    // Assert that the ExpenseList component is rendered
    expect(container).toMatchSnapshot();
});