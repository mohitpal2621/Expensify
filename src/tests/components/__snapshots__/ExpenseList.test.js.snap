// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render expense list with empty expenses 1`] = `
<div>
  <div>
    <h1>
      This is the Expense List
    </h1>
    <p>
      No Expense
    </p>
  </div>
</div>
`;

exports[`Should render expense list with expenses 1`] = `
<div>
  <div>
    <h1>
      This is the Expense List
    </h1>
    <div>
      <a
        href="/edit/1"
      >
        <h3>
          Gum
        </h3>
      </a>
      <p>
        140
         - 
        0
      </p>
    </div>
    <div>
      <a
        href="/edit/2"
      >
        <h3>
          Rent
        </h3>
      </a>
      <p>
        14000.94
         - 
        -345600000
      </p>
    </div>
    <div>
      <a
        href="/edit/3"
      >
        <h3>
          Credit
        </h3>
      </a>
      <p>
        113221.8
         - 
        345600000
      </p>
    </div>
  </div>
</div>
`;
